pkg/backend/filestate/stack.go

```diff
@@ -40,27 +42,49 @@ type Stack interface {
 
 // localStack is a local stack descriptor.
 type localStack struct {
-	ref      *localBackendReference // the stack's reference (qualified name).
-	path     string                 // a path to the stack's checkpoint file on disk.
-	snapshot *deploy.Snapshot       // a snapshot representing the latest deployment state.
-	b        *localBackend          // a pointer to the backend this stack belongs to.
+	ref        *localBackendReference // the stack's reference (qualified name).
+	path       string                 // a path to the stack's checkpoint file on disk.
+	snapshot   *deploy.Snapshot       // a snapshot representing the latest deployment state, allocated on first use.
+	checkpoint *apitype.CheckpointV3  // the raw checkpoint for this stack, deallocated once the snapshot is set.
+	b          *localBackend          // a pointer to the backend this stack belongs to.
 }
 
-func newStack(ref *localBackendReference, path string, snapshot *deploy.Snapshot, b *localBackend) Stack {
+func newStack(ref *localBackendReference, path string, checkpoint *apitype.CheckpointV3, b *localBackend) Stack {
 	contract.Requiref(ref != nil, "ref", "ref was nil")
 
 	return &localStack{
-		ref:      ref,
-		path:     path,
-		snapshot: snapshot,
-		b:        b,
+		ref:        ref,
+		path:       path,
+		checkpoint: checkpoint,
+		b:          b,
 	}
 }
 
 func (s *localStack) Ref() backend.StackReference { return s.ref }
 func (s *localStack) Snapshot(ctx context.Context, secretsProvider secrets.Provider) (*deploy.Snapshot, error) {
-	// TODO: We should delay snapshot reading till here like we do for httpstate, which _also_ means we can
-	// use the secretsProvider passed in rather than assuming DefaultSecretsProvider.
+	if s.snapshot != nil {
```

Comment:
Feel like we should probably consider that for httpstate as well. Also not sure if this should just lazy decode the checkpoint or if we should actually defer reading the checkpoint till here.

Reply:
Given the race condition panics we've seen around concurrent writing of snapshots, it might be nice to err on the side of caution re: locking. 

I don't know enough about the guts here to say for certain, just wanted to offer a reminder of those P1s so you're thinking about it too :)